### PR TITLE
`quantum_rel_entr`, `quantum_cond_entr` && CVXPY QI Marimo demo

### PR DESCRIPTION
**Pull request by aryamanjeendgar**
## Description
This PR implements two extremely important atoms for QI modelling: `quantum_rel_entr` and `quantum_cond_entr`. I also decided to write a Marimo notebook with excerpts from my thesis bringing together a bunch of really nice and cool QI problems that CVXPY is now capable of solving! (the notebook could definitely use an overhaul XD)

## Type of change
- [x] New feature (backwards compatible)
- [ ] New feature (breaking API changes)
- [ ] Bug fix
- [ ] Other (Documentation, CI, ...)

## [Contribution checklist](https://www.cvxpy.org/contributing/index.html#contribution-checklist)
- [ ] Add our license to new files.
- [x] Check that your code adheres to our coding style.
- [x] Write unittests.
- [x] Run the unittests and check that they’re passing.
- [x] Run the benchmarks to make sure your change doesn’t introduce a regression.

### CODE REVIEW

File: cvxpy/tests/test_quantum_rel_entr.py [hunk 1]
```diff
@@ -0,0 +1,241 @@
...
+def randH(n: int):
+    A = np.random.randn(n, n) + 1j * np.random.randn(n, n)
+    return (A + A.conj().T)/2
+
+def randRho(n: int):
+    p = 10 * randH(n)
+    p = (p @ p.conj().T)/np.trace(p @ p.conj().T)
+    return p
+
+class Test_quantum_rel_entr:
```
Comment (SteveDiamond) [inline cvxpy/tests/test_quantum_rel_entr.py:39]: Should be camel case "TestQuantumRelEntr"

File: cvxpy/atoms/quantum_cond_entr.py [hunk 2]
```diff
@@ -0,0 +1,22 @@
+from cvxpy.atoms.affine.kron import kron
+from cvxpy.atoms.affine.partial_trace import partial_trace
+from cvxpy.atoms.quantum_rel_entr import quantum_rel_entr
+
+import numpy as np
+
+
+# if sys == 1
```
Comment (SteveDiamond) [inline cvxpy/atoms/quantum_cond_entr.py:8]: Delete old comments

**Comment by SteveDiamond** [comment]
Thanks @aryamanjeendgar ! You may need to install the precommit https://www.cvxpy.org/contributing/index.html

File: cvxpy/atoms/quantum_cond_entr.py (hunk 2 above)
Comment (aryamanjeendgar) [reply cvxpy/atoms/quantum_cond_entr.py:8]: Whoops, sorry! There's a bunch of them --- will remove them when you're done with the first round of review!

File: cvxpy/atoms/quantum_cond_entr.py [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+#     cvx_optval = -quantum_rel_entr(rho,kron(eye(dim(1)),TrX(rho,sys,dim)));
+# elseif sys == 2
+#     cvx_optval = -quantum_rel_entr(rho,kron(TrX(rho,sys,dim),eye(dim(2))));
+# end
+
+def quantum_cond_entr(rho, dim: list[int], sys=0):
```
Comment (SteveDiamond) [inline cvxpy/atoms/quantum_cond_entr.py:14]: Add types for rho and sys

**Comment by SteveDiamond** [comment]
looks good overall. you can address all the comments then it is ok to merge.

**Comment by SteveDiamond** [comment]
actually you need to fix syntax errors.

File: cvxpy/atoms/quantum_cond_entr.py (hunk 2 above)
Comment (aryamanjeendgar) [reply cvxpy/atoms/quantum_cond_entr.py]: Done!

File: cvxpy/atoms/quantum_cond_entr.py (hunk 3 above)
Comment (aryamanjeendgar) [reply cvxpy/atoms/quantum_cond_entr.py]: Done!

**Comment by aryamanjeendgar** [comment]
@SteveDiamond it looks like the python 3.9 builds are failing since I use a `match-case` in a helper function inside of `test_quantum_rel_entr.py` --- though it's an easy fix, I'm curious if we are targeting 3.9 support for the 1.6 release?

**Comment by SteveDiamond** [comment]
> @SteveDiamond it looks like the python 3.9 builds are failing since I use a `match-case` in a helper function inside of `test_quantum_rel_entr.py` --- though it's an easy fix, I'm curious if we are targeting 3.9 support for the 1.6 release?

We are still targeting 3.9 support.

File: cvxpy/tests/test_quantum_rel_entr.py [hunk 4]
```diff
@@ -184,39 +177,6 @@ def Ic(rho: cp.Variable):
 
         return sth
 
-    @staticmethod
-    def make_test_5():
```
Comment (SteveDiamond) [inline cvxpy/tests/test_quantum_rel_entr.py]: Why are you deleting test 5?

Comment (aryamanjeendgar) [reply cvxpy/tests/test_quantum_rel_entr.py]: This test has kinda weird behaviour across different python versions --- I am comparing solutions against CVXQUAD (Matlab), and for my setup, the problem solves as unbounded but feasible --- whereas on some of the failing test runs I saw, it was outputting the correct CVXQUAD objective value --- even though with `verbose` on, the final primal value of the objective that I am shown is actually the correct objective value, before the solver jumps and outputs an unbounded objective.

I have a feeling the behaviour could be arising from the way I might have implemented some of the additional QI functionalities I needed for this test (like `applychan`)

Comment (SteveDiamond) [reply cvxpy/tests/test_quantum_rel_entr.py]: Ok that's fine.

File: cvxpy/reductions/complex2real/canonicalizers/matrix_canon.py [hunk 5]
```diff
@@ -129,6 +129,28 @@ def von_neumann_entr_canon(expr: von_neumann_entr,
     return canon_expr, None
 
 
+def quantum_rel_entr_canon(expr: quantum_rel_entr,
+                           real_args: List[Union[Expression, None]],
+                           imag_args: List[Union[Expression, None]], real2imag):
+    """
+    TODO: WRITE ME.
```
Comment (SteveDiamond) [inline cvxpy/reductions/complex2real/canonicalizers/matrix_canon.py:136]: Can we put something here?

Comment (aryamanjeendgar) [reply cvxpy/reductions/complex2real/canonicalizers/matrix_canon.py:136]: Do you mean the `TODO`? I think this is just an old-dangling comment, can/should be removed

Comment (SteveDiamond) [reply cvxpy/reductions/complex2real/canonicalizers/matrix_canon.py:136]: But we could add a description of what is happening in the canonicalization.

File: cvxpy/tests/test_quantum_rel_entr.py [hunk 6]
```diff
@@ -0,0 +1,202 @@
...
+    """
+    Test class for `quantum_rel_entr` & `quantum_cond_entr`
+    - All of the reference solutions for the problem come from equivalent
+    CVXQUAD implementations
+    - These problems also show up in a Marimo notebook linked on the CVXPY docs
+    """
+    if 'MOSEK' in cp.installed_solvers():
+        SOLVE_ARGS = {'solver': 'MOSEK', 'verbose': True}
+    else:
+        SOLVE_ARGS = {'solver': 'SCS', 'eps': 1e-6, 'max_iters': 500_000,
```
Comment (SteveDiamond) [inline cvxpy/tests/test_quantum_rel_entr.py:42]: This is too many iterations. Should be at most 10k. It is taking too long in the CI.

Comment (SteveDiamond) [reply cvxpy/tests/test_quantum_rel_entr.py:42]: Clarabel handles SDP so you can use Clarabel.

Comment (SteveDiamond) [reply cvxpy/tests/test_quantum_rel_entr.py]: You need to change the arguments for Clarabel. You can delete eps and max_iters.

File: cvxpy/tests/test_quantum_rel_entr.py [hunk 7]
```diff
@@ -36,10 +37,9 @@ class TestQuantumRelEntr:
     CVXQUAD implementations
     - These problems also show up in a Marimo notebook linked on the CVXPY docs
     """
-    if 'MOSEK' in cp.installed_solvers():
-        SOLVE_ARGS = {'solver': 'MOSEK', 'verbose': True}
-    else:
-        SOLVE_ARGS = {'solver': 'CLARABEL', 'verbose': True}
+    run_full_test_suite = 'MOSEK' in cp.installed_solvers()
+    SOLVE_ARGS = {'solver': 'MOSEK', 'verbose': True} if run_full_test_suite\
```
Comment (SteveDiamond) [inline cvxpy/tests/test_quantum_rel_entr.py:41]: This is not quite right. test_1 and test_3 should use CLARABEL no matter what. test_2 and test_4 should only run if MOSEK is present, but in a very specific way. At the start of test_2 and test_4 add a check

```
if cp.MOSEK in cp.installed_solvers():
  run test
```

Comment (SteveDiamond) [reply cvxpy/tests/test_quantum_rel_entr.py:41]: so the tests should always run no matter what, but test_2 and test_4 do nothing if mosek is not present

Comment (SteveDiamond) [reply cvxpy/tests/test_quantum_rel_entr.py:41]: also @rileyjmurray if you can figure out how to make the tests all work with Clarabel that would be very helpful. but that can be a follow up PR.

Comment (aryamanjeendgar) [reply cvxpy/tests/test_quantum_rel_entr.py:41]: Oh right --- as of now, `test_{2 | 4}` only run when MOSEK is installed (and the entire test suite is run on MOSEK) but if its not found then they are skipped and `test_{1 | 3}` are run with CLARABEL.

Let me make this change really quick

File: cvxpy/atoms/quantum_rel_entr.py [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+            val = -entropy(x)
+            un_normalized = (x_sum * val + np.log(x_sum)*x_sum)
+            return un_normalized
+        r1 = -entropy(w1)
+        r2 = u @ np.log(w2)
+        return (r1 - r2)
+
+    def validate_arguments(self) -> None:
+        """Verify that the argument A is PSD.
+        """
```
Comment (rileyjmurray) [inline cvxpy/atoms/quantum_rel_entr.py:64]: I believe this constraint will be enforced by canonicalization, so no comment to this effect is necessary.

File: cvxpy/atoms/quantum_rel_entr.py [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+        N = self.args[0]
+        if N.size > 1:
+            if N.ndim != 2 or N.shape[0] != N.shape[1]:
+                raise ValueError('Argument must be a square matrix.')
```
Comment (rileyjmurray) [inline cvxpy/atoms/quantum_rel_entr.py:68]: Replace this with checks that ``args[0].is_hermitian()`` and ``args[1].is_hermitian()``.

File: cvxpy/atoms/quantum_rel_entr.py [hunk 10]
```diff
@@ -0,0 +1,119 @@
...
+
+        Matrix expressions are vectorized, so the gradient is a matrix.
+
+        Args:
+            values: A list of numeric values for the arguments.
+
+        Returns:
+            A list of SciPy CSC sparse matrices or None.
+        """
+        raise ValueError()
```
Comment (rileyjmurray) [inline cvxpy/atoms/quantum_rel_entr.py:114]: Raise NotImplementedError

File: cvxpy/atoms/quantum_rel_entr.py [hunk 11]
```diff
@@ -0,0 +1,119 @@
...
+
+    def _domain(self) -> List[Constraint]:
+        """Returns constraints describing the domain of the node.
+        """
+        return [self.args[0] >> 0]
```
Comment (rileyjmurray) [inline cvxpy/atoms/quantum_rel_entr.py:119]: We also need ``self.args[1] >> 0``.

File: cvxpy/tests/test_quantum_rel_entr.py [hunk 12]
```diff
@@ -0,0 +1,205 @@
...
+        obj_pair = (obj, obj_expect)
+
+        cons1 = p1 >= 0
+        cons2 = p2 >= 0
+        cons3 = p1 + p2 == 1
+        cons_pair = [(cons1, None), (cons2, None), (cons3, None)]
+
+        sth = STH.SolverTestHelper(obj_pair, var_pairs, cons_pair)
+
+        return sth
```
Comment (rileyjmurray) [inline cvxpy/tests/test_quantum_rel_entr.py:135]: @aryamanjeendgar am I misunderstanding, or does this not rely on quantum relative entropy? Seems this test belongs in test_von_neumann_entr.py.

File: cvxpy/reductions/complex2real/canonicalizers/matrix_canon.py [hunk 13]
```diff
@@ -129,6 +129,28 @@ def von_neumann_entr_canon(expr: von_neumann_entr,
...
+    assert all(ra is not None for ra in real_args)
+    expanded_X = expand_complex(real_args[0], imag_args[0])
+    expanded_Y = expand_complex(real_args[1], imag_args[1])
+    canon_expr = expr.copy([expanded_X, expanded_Y])
+    canon_expr = canon_expr / 2
+    # ^ I'm guessing this is the correct normalization, since the
+    #   spectra of (expanded_X, expanded_Y) is just the spectra
+    #   of (X, Y) repeated twice. We'll need a unit test that
+    #   checks correctness relative to CVXQUAD output for some test
+    #   problem. -- Riley
```
Comment (rileyjmurray) [inline cvxpy/reductions/complex2real/canonicalizers/matrix_canon.py:150]: This comment can be removed. We've established the correct scaling with ``TestQuantumRelEntr::test_2``.

File: cvxpy/reductions/complex2real/canonicalizers/matrix_canon.py (hunk 5 above)
Comment (rileyjmurray) [reply cvxpy/reductions/complex2real/canonicalizers/matrix_canon.py:136]: This text will suffice
```
    """Transform Hermitian input for quantum_rel_entr into equivalent
    symmetric input for quantum_rel_entr.
    """
```

File: cvxpy/tests/test_quantum_rel_entr.py (hunk 12 above)
Comment (rileyjmurray) [inline cvxpy/tests/test_quantum_rel_entr.py:135]: @aryamanjeendgar am I misunderstanding, or does this not rely on quantum relative entropy? Seems this test belongs in test_von_neumann_entr.py.

**Review by rileyjmurray — CHANGES_REQUESTED** [review]
I left a few minor comments. Should be good to merge once they're addressed.

File: cvxpy/tests/test_quantum_rel_entr.py (hunk 12 above)
Comment (aryamanjeendgar) [reply cvxpy/tests/test_quantum_rel_entr.py]: You're right! I put all of the problems from the Hamza paper in the notebook and didn't check --- thank you for the catch! This has been added to the `Test_von_neumann_entr` test-suite!

**Comment by aryamanjeendgar** [comment]
@rileyjmurray Done! The latest commit should address all comments!

**Comment by Transurgeon** [comment]
Is this ready to merge? I'd like to do the release today if possible :)

**Comment by aryamanjeendgar** [comment]
> Is this ready to merge? I'd like to do the release today if possible :)
I believe so --- @rileyjmurray, is there anything else you'd want me to address?

**Comment by rileyjmurray** [comment]
Aryaman's latest changes look good to me. Ready to merge!

**Review by Transurgeon — APPROVED** [review]

**Pull request merged**